### DISCUSSION TITLE
How to improve the performance of facerecognition-external-model (for nextcloud-aio)

### DISCUSSION THREAD

**Discussion by luzfcb** [matiasdelellis/facerecognition#716] [Q&A]
I started testing Nextcloud AIO v7.10.0 Beta. Is there any way to improve performance of facerecognition-external-model?
For me, it analyzes 1 image every 16 seconds.
My server runs on Ryzen 5600G, 32 GB of RAM, and Nvidia GTX 1650 GPU, so I suppose it's more than enough to have a very good performance.

**Comment by luzfcb** [comment]
from https://github.com/goodspb/pdlib/issues/56#issuecomment-1884894558 by @matiasdelellis 

Hi @luzfcb

> My server runs on Ryzen 5600G, 32 GB of RAM, and Nvidia GTX 1650 GPU, so I suppose it's more than enough to have a very good performance.

Mainly it's great that you have an Nvidia GTX 1650 😬 , but the docker container is not compiled to use cuda. You should install CUDA on your host, and build the container to make use of it. 😉
It will be much faster. 😬
It shouldn't be complicated, but I can't help you because I never had a video card. 😞

On the other hand, you can reduce the size of temporary images on settings. By default this container is in 4Mpx, and with 2Mps you will get it will be 50% faster and you will get practically the same result.

I left 4mpx as the default because the larger the images, the more likely you are to find all your faces, but in practice, you don't get more than 10% new faces.

Finally no matter how long the process takes, it is designed to do the job progressively. We try not to take shortcuts but do better.

p.s: Not even Google gives you the results so quickly when you upload many images... 😅
Let's be pacient.. 😉

**Comment by szaimen** [comment]
I think one idea would be parallelization of the analyzation? Currrently it looks like only one image is analyzed at a time... WDYT @matiasdelellis ?

**Reply by matiasdelellis** [reply]
Currently there are a global lock that prevents the execution of more than one command. I always expected that there would be some standard parallelism API in PHP, but none fit what I would like to use. 

MM... pthreads is deprecated and Fibers is not real threads.

Now I'm thinking about using an approach similar to Memories. This is using locks on each scanned file which still doesn't convince me, but it's such a simple approach that it can work.

**Reply by szaimen** [reply]
> Now I'm thinking about using an approach similar to Memories. This is using locks on each scanned file which still doesn't convince me, but it's such a simple approach that it can work

Sounds good to me :)

**Reply by matiasdelellis** [reply]
We have something.. 🤪 
![imagen](https://github.com/matiasdelellis/facerecognition/assets/733715/e6fe68bb-1800-45f8-9532-902cb9916039)

https://github.com/matiasdelellis/facerecognition/issues/237#issuecomment-1888988522

Note that unfortunately the external model is still single threaded, and will probably fail with timeouts. It needs to be investigated. 😞

**Reply by matiasdelellis** [reply]
Small update..
We don't need to change the external model since we can make a pool or load balancer with gunicorn very easily.

Just call:
>$ gunicorn -w 4 'facerecognition-external-model:app' --threads 4 --access-logfile -

However I am looking for something as simple as this to run the `occ face:background_job --analyze-mode` command several times and wait for it to finish, since all of them must finish before running `occ face:background_job --cluster-mode`

🤔 ps: I already saw `parallel`, I know `wait`, `$!`, etc, but nothing is as simple as I want. 😞

**Reply by matiasdelellis** [reply]
Well, 
I never commented this, it's definitely not elegant but we have to use something like this.
```
#!/bin/bash
echo "Synchronizing user files..."
php occ face:background_job -u user --sync-mode
echo "Done"

echo "Analyzing user files..."
for i in {1..4}; do
    php occ face:background_job -u user --analyze-mode &
    pids[${i}]=$!
done

for pid in ${pids[*]}; do
    wait $pid
done
echo "Done"

echo "Calculating user face clusters..."
php occ face:background_job -u user --cluster-mode
echo "Done"
```

**Reply by doppelgrau** [reply]
And it seems to work, see the pull request for the docker image. Reducing for me the import of one of my family members to a somewhat manageable time.

**Reply by MoradDerhourhi** [reply]
script worked fine for me thanks🙏🙏

**Comment by wojciechczyz** [comment]
To speed up processin, I am following this guide and I am running 4 background jobs on the machine with 4 CPU cores. Very frecuently they keep having problem resolving external model docker container. Why is that?

Faces found: 0. Image will be skipped because of the following error: Could not resolve host: nextcloud-aio-facerecognition
        Processing image /mnt/external_storage/Photos/PhotosExport/IMG_7650.JPG

        Faces found: 0. Image will be skipped because of the following error: Could not resolve host: nextcloud-aio-facerecognition
        Processing image /mnt/external_storage/Photos/PhotosExport/20210412_185714.jpg

        Processing image /mnt/external_storage/Photos/PhotosExport/DSC_1351.JPG
        Faces found: 1
        Processing image /mnt/external_storage/Photos/PhotosExport/IMG_9096 (2).JPG
        Faces found: 3
        Processing image /mnt/external_storage/Photos/PhotosExport/20210222_133830_IMG_2223.HEIC

**Reply by szaimen** [reply]
You could try if AIO v9.7.0 beta resolves the connection problems. See https://github.com/nextcloud/all-in-one?tab=readme-ov-file#how-to-switch-the-channel

**Comment by EngTurtle** [comment]
Hey, if you're willing to start analysis from scratch I've built an external model container that uses the onnx framework and the InsightFace models. Using OpenVINO acceleration my intel N100 server only takes about _250ms_ to _1s_ to scan an image. Following immich's ML container example I've also included a CUDA accelerated version, but haven't had the chance to test it yet.

https://github.com/matiasdelellis/facerecognition/discussions/826
